extensions/guacamole-auth-radius/src/main/java/org/apache/guacamole/auth/radius/AuthenticationProviderService.java

```diff
@@ -0,0 +1,223 @@
+                logger.error("Cannot configure RADIUS server: {}", e.getMessage());
+                logger.debug("Error configuring RADIUS server.", e);
+                throw new GuacamoleInvalidCredentialsException("Authentication error.", CredentialsInfo.USERNAME_PASSWORD);
+            }
+
+            // If configure fails, permission to login is denied
+            if (radPack == null) {
+                logger.debug("Nothing in the RADIUS packet.");
+                throw new GuacamoleInvalidCredentialsException("Authentication error.", CredentialsInfo.USERNAME_PASSWORD);
+            }
+
+            // If we get back an AccessReject packet, login is denied.
+            else if (radPack instanceof AccessReject) {
+                logger.debug("Login has been rejected by RADIUS server.");
+                throw new GuacamoleInvalidCredentialsException("Permission denied.", CredentialsInfo.USERNAME_PASSWORD);
+            }
+
+            // If we receive an AccessChallenge package, the server needs more information -
+            // We create a new form/field with the challenge message.
+            else if (radPack instanceof AccessChallenge) {
+
+                // Try to get the state attribute - if it's not there, we have a problem
+                RadiusAttribute stateAttr = radPack.findAttribute(Attr_State.TYPE);
+                if (stateAttr == null) {
+                    logger.error("Something went wrong, state attribute not present.");
+                    logger.debug("State Attribute turned up null, which shouldn't happen in AccessChallenge.");
+                    throw new GuacamoleInvalidCredentialsException("Authentication error.", CredentialsInfo.USERNAME_PASSWORD);
+                }
+
+                // We need to get the reply message so we know what to ask the user
+                RadiusAttribute replyAttr = radPack.findAttribute(Attr_ReplyMessage.TYPE);
+                if (replyAttr == null) {
+                    logger.error("No reply message received from the server.");
+                    logger.debug("Expecting a Attr_ReplyMessage attribute on this packet, and did not get one.");
+                    throw new GuacamoleInvalidCredentialsException("Authentication error.", CredentialsInfo.USERNAME_PASSWORD);
+                }
+
+                // We have the required attributes - convert to strings and then generate the additional login box/field
+                String replyMsg = replyAttr.toString();
+                String radiusState = new String(stateAttr.getValue().getBytes());
+                Field radiusResponseField = new RadiusChallengeResponseField(replyMsg);
+                Field radiusStateField = new RadiusStateField(radiusState);
+                CredentialsInfo expectedCredentials = new CredentialsInfo(Arrays.asList(radiusResponseField,radiusStateField));
+                throw new GuacamoleInsufficientCredentialsException("LOGIN.INFO_RADIUS_ADDL_REQUIRED", expectedCredentials);
```

Comment:
The challenge from the server is exposed to the user through the radiusResponseField - this field is initialized with the replyMsg (which is replyAttr.toString()), which is the challenge from the server, and passed through in the expectedCredentials section.  This shows up in the form presented to the user.